app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,598 @@
+  public static final int REQUEST_ENABLE_WIFI_P2P = 1;
+  public static final int REQUEST_ENABLE_LOCATION_SERVICES = 2;
+  private static final int PERMISSION_REQUEST_CODE_COARSE_LOCATION = 1;
+  private static final int PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS = 2;
+
+  public static int PEER_HANDSHAKE_PORT = 8009;
+  public static int FILE_TRANSFER_PORT = 8008;
+
+  @Inject SharedPreferenceUtil sharedPreferenceUtil;
+  @Inject AlertDialogShower alertDialogShower;
+
+  @BindView(R.id.toolbar_local_file_transfer) Toolbar actionBar;
+  @BindView(R.id.text_view_device_name) TextView deviceName;
+  @BindView(R.id.progress_bar_searching_peers) ProgressBar searchingPeersProgressBar;
+  @BindView(R.id.list_peer_devices) ListView listViewPeerDevices;
+  @BindView(R.id.text_view_empty_peer_list) TextView textViewPeerDevices;
+  @BindView(R.id.recycler_view_transfer_files) RecyclerView filesRecyclerView;
+
+  private ArrayList<Uri> fileUriArrayList; // For sender device, stores uris of the files
+  public @NonNull Boolean fileSendingDevice = false;// Whether the device is the file sender or not
+
+  public @NonNull WifiDirectManager wifiDirectManager = new WifiDirectManager(this);
+
+  private int totalFilesForTransfer = -1;
+  private int filesSent = 0;          // Count of number of files transferred until now
+  private ArrayList<FileItem> filesToSend = new ArrayList<>();
+  private FileListAdapter fileListAdapter;
+
+  private List<WifiP2pDevice> peerDevices = new ArrayList<WifiP2pDevice>();
+  private InetAddress selectedPeerDeviceInetAddress;
+  private InetAddress fileReceiverDeviceAddress;  // IP address of the file receiving device
+  private boolean fileTransferStarted = false;
+
+  private PeerGroupHandshakeAsyncTask peerGroupHandshakeAsyncTask;
+  private SenderDeviceAsyncTask[] senderDeviceAsyncTaskArray;
+  private ReceiverDeviceAsyncTask receiverDeviceAsyncTask;
+
+  @Override
+  protected void onCreate(@Nullable Bundle savedInstanceState) {
+    KiwixApplication.getApplicationComponent().activityComponent()
+        .activity(this)
+        .build()
+        .inject(this);
+    setTheme(sharedPreferenceUtil.nightMode() ? R.style.Theme_AppCompat_DayNight_NoActionBar : R.style.AppTheme);
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_local_file_transfer);
+    ButterKnife.bind(this);
+
+    /*
+     * Presence of file Uris decides whether the device with the activity open is a sender or receiver:
+     * - On the sender device, this activity is started from the app chooser post selection
+     * of files to share in the Library
+     * - On the receiver device, the activity is started directly from within the 'Get Content'
+     * activity, without any file Uris
+     * */
+    Intent filesIntent = getIntent();
+    fileUriArrayList = filesIntent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);
+    if (fileUriArrayList != null && fileUriArrayList.size() > 0) {
+      setDeviceAsFileSender();
```

Comment:
a method that starts with set and takes no parameter is very suspicious.
This is the only place you set fileSendingDevice so it would be just as clean to write
```
fileSendingDevice = fileUriArrayList != null && fileUriArrayList.size() > 0
```
or
```
fileSendingDevice = hasFileList(getIntent().getParcelableArrayListExtra(Intent.EXTRA_STREAM))
```

Also the naming is inconsistent `setDeviceAsFileSender/fileSendingDevice/isFileSender`, I think `isFileSender` is best. Starting a Boolean with a question word is encouraged because it makes for much nicer sentences in if statements (See section [3.13](https://petroware.no/javastyle.html)).


Further still I have a problem with the public `isFileSender()` method. It is used in one place which is PeerGroupHandshakeAsyncTask. At the time of construction you already know if it is a fileSender or not so really this class should be 2 classes instead of branching logic based on a method call to a dependent class. If the aim was to reuse common logic than this is achieved through a common base class (usually not the best) or through composition and sharing an object that performs those similar duties between them, HandshakeInitiator maybe.

Infact most if not all the things you query from LoclFileTransferActivity to PeerGroupHandshakeAsyncTask should be passed in to its constructor

Reply:
I have removed method `setDeviceAsFileSender()` and corrected the variable name to `isFileSender`

Regarding the use of a single class instead of two despite branching the code based on whether it's the wifi-direct peer group's owner or not (this is independent of being sender or receiver), I did it in an attempt to keep related code as close together as possible - which now I gather violates design principles.

I am thinking of creating a `HandshakeAsyncTask` interface with the method `exchangeFileTransferMetadata()`, and then subclass it to create a `GroupOwnerHandshakeTask` (with code for creating the server socket), and `NonGroupOwnerHandshakeTask` (with code for creating the client socket).

Though, since a device being sender or receiver is not dependant on whether the device is group owner or not, the `exchangeFileTransferMetadata()` method will still have branching

Reply:
@macgills I think I'll look into this when updating the code for Rx in #1351? Might end up redoing things otherwise

Reply:
sure thing